torch_geometric/nn/conv/hgt_conv.py

```diff
@@ -143,40 +178,105 @@ def forward(
         """
         H, D = self.heads, self.out_channels // self.heads
 
-        k_dict, q_dict, v_dict, out_dict = {}, {}, {}, {}
+        k_dict, q_dict, v_dict = {}, {}, {}
+        out_dict = defaultdict(list)
+
+        # parallelize over node-types
+        # compute K, Q, V over node-types
+        k_dict = {
+            node_type: k_j.view(-1, H, D)
+            for node_type, k_j in self.k_lin(x_dict).items()
+        }
+        q_dict = {
+            node_type: q_j.view(-1, H, D)
+            for node_type, q_j in self.q_lin(x_dict).items()
+        }
+        v_dict = {
+            node_type: v_j.view(-1, H, D)
+            for node_type, v_j in self.v_lin(x_dict).items()
+        }
+
+        # parallelize over edge-types
+        a_rels = [
+            self.a_rel['__'.join(edge_type)] for edge_type in self.edge_types
+        ]
+        m_rels = [
+            self.m_rel['__'.join(edge_type)] for edge_type in self.edge_types
+        ]
+
+        k_ins = []
+        v_ins = []
+        count = 0
+        trans_ptr = [count]
+        for src_type in self.src_types:
+            k_src = k_dict[src_type]
+            v_src = v_dict[src_type]
+            k_ins.append(k_src.view(-1, D))
+            v_ins.append(v_src.view(-1, D))
+            for i in range(H):
+                count += k_src.size(0)
+                trans_ptr.append(count)
+        trans_ptr = torch.tensor(trans_ptr)
+        print("Trans", trans_ptr.shape, trans_ptr)
+        a_rel, m_rel = torch.cat(a_rels), torch.cat(m_rels)
+        k_out = pyg_lib.ops.segment_matmul(torch.cat(k_ins), trans_ptr,
+                                           a_rel).view(-1, H, D)
+
+        v_out = pyg_lib.ops.segment_matmul(torch.cat(v_ins), trans_ptr,
+                                           m_rel).view(-1, H, D)
+        print("K_out", k_out.shape)
+        node_slices = make_node_slices({
+            n_type: k_out[trans_ptr[i]:trans_ptr[i + 1]].shape[0]
+            for i, n_type in enumerate(self.node_types)
+        })
+        print("node_slices", node_slices)
+        # combine edge_index dict into single tensor
+        q_list = []
+        p_rels = []
+        edge_index_list = []
+        for e_type in self.edge_types:
+            indices = edge_index_dict[e_type]
+            # (TODO) Add support for SparseTensor w/o converting?
+            convert = isinstance(indices, SparseTensor)
+            if convert:
+                # convert to COO
+                dst, src, _ = indices.coo()
+                indices = torch.cat((src.view(1, -1), dst.view(1, -1)))
+            edge_index_list.append(
+                offset_edge_idx(node_slices, e_type, indices))
+            q_list.append(q_dict[e_type[-1]])
+            p_rels.append(self.p_rel['__'.join(e_type)].view(-1, 1))
```

Comment:
I have a concern with `node_slices` and how the `offset_edge_idx` is being applied. I thought it would be best to share an example to illustrate the issue I see: 

I added a few print stmts to the code (we'll remove them before merging) and ran the below example:
```
import torch
from torch_geometric.nn import HGTConv

x_dict = {
    'author': torch.randn(4, 16),
    'paper': torch.randn(6, 16),
}

index1 = torch.randint(0, 4, (20, ), dtype=torch.long)
index2 = torch.randint(0, 6, (20, ), dtype=torch.long)

edge_index_dict = {
    ('author', 'writes', 'paper'): torch.stack([index1, index2]),
    ('paper', 'written_by', 'author'): torch.stack([index2, index1]),
    ('paper', 'liked_by', 'author'): torch.stack([index2, index1]),
}
metadata = (list(x_dict.keys()), list(edge_index_dict.keys()))
conv = HGTConv(16, 16, metadata, heads=2)
out_dict1 = conv(x_dict, edge_index_dict)
```
The output is as follows:
```
Trans torch.Size([7]) tensor([ 0,  4,  8, 14, 20, 26, 32])
K_out torch.Size([16, 2, 8])
node_slices {'author': (0, 4), 'paper': (4, 8)}
Q torch.Size([14, 2, 8])
```
In the above case the `node_slices` I believe are for `k_out` but since `k_out` has node types repeated, since `src_types` has repeated node types, I don't see how these `node_slices` can be used to offset the `edge_index` appropriately. 

Again I might be missing something, since the code is kinda complex.

Reply:
this code for the node_slices is exactly copied from `to_homogeneous_edge_index` which was written by @rusty1s 

from my understanding this code has been working for cases even when there are multiple of the same source types. the node slices are applied as offsets to the edge indices w/:
```
def offset_edge_idx(
    node_slices: Dict[NodeType, Tuple[int, int]],
    edge_type: EdgeType,
    edge_index: Tensor,
) -> Tuple[int, int]:
    src, _, dst = edge_type
    offset = [[node_slices[src][0]], [node_slices[dst][0]]]
    offset = torch.tensor(offset, device=edge_index.device)
    return edge_index + offset
```

im not sure if i see any issue. @rusty1s plz advise

Reply:
Also discussed with @wsad1. I haven't looked to closely yet, but I think he is right. `K_out` seems to be of shape `[16, 2, 8]`, where `16 == 4(author)+6(paper)+6(paper)`, so we cannot use `node_slices` here. We would need an edge-type dependent `node_slices` vector instead.

Reply:
Thanks for explaining that. 
Yeah we need `node_slices` per edge type, further the offset for edge_index[0] should depend on slices in `k_out` and for edge_index[1] it should depend on slices in `q`. Happy to update this if that helps.

Reply:
> Happy to update the PR, if that helps.

@wsad1  that would be appreciated if possible. thanks!